File: implementation/scipamato/core/persistence-jooq/src/main/java/ch/difty/scipamato/core/persistence/search/SearchOrderFilterConditionMapper.java
package ch.difty.scipamato.core.persistence.search;

import static ch.difty.scipamato.core.db.tables.SearchOrder.SEARCH_ORDER;

import java.util.List;

import org.jetbrains.annotations.NotNull;
import org.jooq.Condition;
import org.jooq.impl.DSL;

import ch.difty.scipamato.common.persistence.AbstractFilterConditionMapper;
import ch.difty.scipamato.common.persistence.FilterConditionMapper;
import ch.difty.scipamato.core.entity.search.SearchOrderFilter;

/**
 * Mapper turning the provider {@link SearchOrderFilter} into a jOOQ
 * {@link Condition}.
 *
 * @author u.joss
 */
@FilterConditionMapper
public class SearchOrderFilterConditionMapper extends AbstractFilterConditionMapper<SearchOrderFilter> {

    @Override
    public void map(@NotNull final SearchOrderFilter filter, @NotNull final List<Condition> conditions) {
        if (filter.getOwnerIncludingGlobal() != null) {
            conditions.add(
                DSL.or(SEARCH_ORDER.OWNER.equal(filter.getOwnerIncludingGlobal()), SEARCH_ORDER.GLOBAL.equal(true)));
        } else {
            if (filter.getNameMask() != null) {
                conditions.add(SEARCH_ORDER.NAME
                    .lower()
                    .contains(filter
                        .getNameMask()
                        .toLowerCase()));
            }
            if (filter.getOwner() != null) {
                conditions.add(SEARCH_ORDER.OWNER.equal(filter.getOwner()));
            }

            if (filter.getGlobal() != null) {
                conditions.add(SEARCH_ORDER.GLOBAL.equal(filter.getGlobal()));
            }
        }
    }
}

File: core/core-persistence-jooq/src/main/java/ch/difty/scipamato/core/persistence/search/SearchOrderFilterConditionMapper.kt
package ch.difty.scipamato.core.persistence.search

import ch.difty.scipamato.common.persistence.AbstractFilterConditionMapper
import ch.difty.scipamato.common.persistence.FilterConditionMapper
import ch.difty.scipamato.core.db.tables.SearchOrder
import ch.difty.scipamato.core.entity.search.SearchOrderFilter
import org.jooq.Condition
import org.jooq.impl.DSL

/**
 * Mapper turning the provider [SearchOrderFilter] into a jOOQ [Condition].
 */
@FilterConditionMapper
class SearchOrderFilterConditionMapper : AbstractFilterConditionMapper<SearchOrderFilter>() {

    override fun internalMap(filter: SearchOrderFilter): List<Condition> {
        val conditions = mutableListOf<Condition>()
        if (filter.ownerIncludingGlobal != null) {
            conditions.add(
                DSL.or(
                    SearchOrder.SEARCH_ORDER.OWNER.equal(filter.ownerIncludingGlobal),
                    SearchOrder.SEARCH_ORDER.GLOBAL.equal(true)
                )
            )
        } else {
            filter.nameMask?.let { mask ->
                conditions.add(
                    DSL.lower(SearchOrder.SEARCH_ORDER.NAME).contains(mask.toLowerCase())
                )
            }
            filter.owner?.let { conditions.add(SearchOrder.SEARCH_ORDER.OWNER.equal(it)) }
            filter.global?.let { conditions.add(SearchOrder.SEARCH_ORDER.GLOBAL.equal(it)) }
        }
        return conditions
    }
}
